TINKERPOP-1730 Gremlin .NET: add support for GraphSON3

jorgebay opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-1730

FlorianHockmann commented on gremlin-dotnet/src/Gremlin.Net/Driver/Messages/ResponseResult.cs:
```diff
@@ -23,13 +23,14 @@
     {
         [JsonProperty(PropertyName = "data")]
-        public List<T> Data { get; set; }
+        public JToken Data { get; set; }
```
You can completely remove the type parameter `T` from the class when you change the type here to `JToken`.

jorgebay commented on gremlin-dotnet/test/Gremlin.Net.IntegrationTest/Process/Traversal/DriverRemoteConnection/GraphTraversalTests.cs:
```diff
@@ -102,9 +102,9 @@ public void g_V_HasXname_markoX_ValueMap_Next()
             var g = graph.Traversal().WithRemote(connection);
 
-            var receivedValueMap = g.V().Has("name", "marko").ValueMap<string, object>().Next();
+            var receivedValueMap = g.V().Has("name", "marko").ValueMap<object, object>().Next();
```
The serialization format for maps changed (see JIRA tickets 1427 and 1658), the result from this traversal was originally a js object:

```js
{ "marko": { "@type": "g:Int32", "@value": 29 } }
```

And changed in GraphSON3 to a type "g:Map" for which we use `IDictionary<object, object>`.

I've started a discussion on the mailing list regarding collections child types: https://lists.apache.org/thread.html/3918353aaa63aa07b69214da24fa7aa0760004227fc57fa2b3bcae86@%3Cdev.tinkerpop.apache.org%3E

FlorianHockmann replied:
Ah ok, I saw the thread on the mailing list, but somehow I didn't draw the connection to this change.

jorgebay commented:
Thanks @FlorianHockmann for the feedback.
I've addressed the issues you mentioned.

About `ValueMap<TKey, TValue>()`, we can continue the discussion on the mailing list: https://lists.apache.org/thread.html/3918353aaa63aa07b69214da24fa7aa0760004227fc57fa2b3bcae86@%3Cdev.tinkerpop.apache.org%3E

FlorianHockmann commented:
VOTE: +1

spmallette commented:
Unfortunately, you'll need to rebase because the Gremlin Server test server script moved out of the pom.xml, but your changes are easy enough.

All tests pass with `docker/build.sh -t -n -i`

VOTE +1

jorgebay commented:
Rebased onto master, tests pass with `mvn clean install -P gremlin-dotnet`.

VOTE +1

I'll merge it once CI jobs finish.
